sys/net/gnrc/netreg/gnrc_netreg.c

```diff
@@ -33,12 +34,98 @@
 /* The registry as lookup table by gnrc_nettype_t */
 static gnrc_netreg_entry_t *netreg[GNRC_NETTYPE_NUMOF];
 
+/** Held while accessing _lock_counter, and also while the exclusive lock is held */
+static mutex_t _lock_for_counter = MUTEX_INIT;
+/** Number of shared locks on netreg. Saturating arithmetic is used; if this
+ * reaches UINT_MAX, the lock will never be freed again. This is likely
+ * accurate, given that it will only happen when the lock was leaked, so it
+ * can't be freed any more anyway.
+ *
+ * This can be accessed only when _lock_for_counter is held; an alternative
+ * implementation with atomics would likely be possible, but for the case of "I
+ * can't do that right now" it would still need a mutex. But that's
+ * optimization that could be done without changing the public API.
+ * */
+static unsigned int _lock_counter = 0;
+/** Held while the netreg lists are being read. This is what the exclusive
+ * users block on if they can't grab the exclusive lock right away.
+ *
+ * This is largely used as a boolean flag (is locked / is unlocked) that is
+ * modified while the _lock_for_counter is held, and will not block because it
+ * is synchronized to _lock_counter being 0. It is only ever accessed outside
+ * _lock_for_counter when waiting for the counter to reach 0, and even then is
+ * released immediately to avoid deadlocks. (Instead, the exclusive acquisition
+ * tries to acquire the _lock_for_counter, and if that fails it queues up again
+ * after the reader that just snatched it).
+ * */
+static mutex_t _lock_wait_exclusive = MUTEX_INIT;
+
 void gnrc_netreg_init(void)
 {
     /* set all pointers in registry to NULL */
     memset(netreg, 0, GNRC_NETTYPE_NUMOF * sizeof(gnrc_netreg_entry_t *));
 }
 
+void gnrc_netreg_acquire_shared(void) {
+    mutex_lock(&_lock_for_counter);
+    if (_lock_counter == 0) {
+        /* At most, this blocks for the very short time until
+         * _gnrc_netreg_acquire_exclusive returns it immediately */
+        mutex_lock(&_lock_wait_exclusive);
+    }
+    if (_lock_counter != UINT_MAX) {
+        _lock_counter += 1;
+    }
+    mutex_unlock(&_lock_for_counter);
+}
+
+void gnrc_netreg_release_shared(void) {
+    mutex_lock(&_lock_for_counter);
+    if (_lock_counter != UINT_MAX) {
+        _lock_counter -= 1;
+    }
```

Comment:
So if we reach `UINT_MAX` for some reason, it will just stay there?

Reply:
Mhh another way to reach this is e.g. to accidentally (or maliciousl?) call `release` when `_lock_counter` is currently 0. How about instead adding a check if `_lock_counter` is already 0 here (to prevent an under-flow) and have an assert that `_lock_counter < UINT_MAX` in acquire instead, so a user that locks over and over again is warned by a failed assertion rather than wondering about (and debugging) a deadlock.

Reply:
The underflow is a good check. The overflow ... I can do that, but I'm not convinced it's better. If the user leaks locks and does require exclusive access, they'll see a deadlock soon, and won't wait 65k packages to see the counter assertion fail. OTOH, if they do not require exclusive access anyway, this would make a device that could happily continue working suddenly break rather than continue with a minor error.

Reply:
Underflow is detected now, please let me know if you're happy with the overflow explanation.

Reply:
Ok, I am fine, but still think it would make a debugging session on the hunt for a deadlock much faster, if there is a failed assertion, even if it takes a while to reach it ;-).